File: src/main/kotlin/linked_list/split_list.kt
class Solution {
    fun splitListToParts(head: ListNode?, k: Int): Array<ListNode?> {
        if(head == null)return arrayOf<ListNode?>(null)
        var cnt = 0
        var h = head
        while(h != null){
            h = h.next
            cnt++
        }       
        var i = 0
        h = head
        val res = Array<ListNode?>(k){null}
        if(cnt <= k){
            while(h != null){
                res[i] = h
                h = h.next
                res[i]?.next = null
                i++
            }
        }
        val size = cnt / k
        val arr = IntArray(size){size}
        var r = cnt % k
        i = 0
        while(r > 0){
            arr[i++] += 1
            r--
        }
        h = head
        var it = 0
        for(i in arr){
            var j = i
            val hd = h
            while(j > 1){
                h = h?.next
                j--
            }
            val t = h
            h = h?.next
            t?.next = null
            res[it++] = hd
        }
        return res
    }
}

File: src/main/kotlin/linked_list/split_list.py
class ListNode:
    def __init__(self, val: int, left = None):
        self.val = val
        self.left = left


class Solution:
    def splitListToParts(self, head: Optional[ListNode], k: int) -> List[Optional[ListNode]]:
        if head is None:
            return [[head]]
        res = [None] * k
        cnt = 0
        h = head
        while h:
            h = h.next
            cnt += 1
        h = head
        i = 0
        if cnt <= k:
            while h:
                tmp = h
                res[i] = tmp
                h = h.next
                res[i].next = None
                i++
            return res
        lst = [cnt // k for i in range(k)]
        i, indx = cnt % k, 0
        while i > 0:
            lst[indx++] += 1
            i--
        indx = 0
        for i in lst:
            c = i
            h = head
            while h and c > 1:
                h = h.next
                c--
            res[indx++] = head
            head = h.next
            h.next = None
        return res
        